File: apps/web/src/types/weather.ts
/**
 * Canonical TypeScript types for Weather & Wind Intelligence (Phase 1 & 2).
 */

export type DataStatus = "LIVE" | "CACHED" | "UNAVAILABLE";
export type DataQuality = "LIVE" | "CACHED" | "FALLBACK" | "UNAVAILABLE";
export type WindState = "CALM" | "LIGHT" | "MODERATE" | "FRESH" | "STRONG" | "GALE";

export interface Coordinate {
  latitude: number;
  longitude: number;
}

export interface WindVector {
  speed_ms: number;
  direction_from_deg: number;
  direction_from_label: string;
  direction_to_deg: number;
  downwind_direction_label: string;
  gust_ms: number | null;
  u_ms: number;
  v_ms: number;
  is_calm: boolean;
  wind_state: WindState;
}

export interface AtmosphereData {
  temperature_c: number;
  relative_humidity_pct: number;
  surface_pressure_hpa: number | null;
  precipitation_mm: number | null;
  cloud_cover_pct: number | null;
  boundary_layer_height_m: number | null;
  soil_moisture_m3_m3: number | null;
}

export interface WeatherForecastPoint {
  forecast_time: string;
  horizon_hours: number;
  atmosphere: AtmosphereData;
  wind: WindVector;
}

export interface WeatherProviderInfo {
  name: string;
  model?: string | null;
}

export interface WeatherResponse {
  location: Coordinate;
  observed_at: string;
  retrieved_at: string;
  data_status: DataStatus;
  data_quality: DataQuality;
  atmosphere: AtmosphereData;
  wind: WindVector;
  forecast: WeatherForecastPoint[];
  provider: WeatherProviderInfo;
}

export interface EventWeatherResponse {
  event_id: string;
  weather: WeatherResponse;
  enriched_at: string;
}

File: services/api/schemas/weather.py
"""Pydantic API schemas for Weather & Wind endpoints (Phase 1 & 2)."""

from pydantic import BaseModel, Field

from packages.schemas.common import Coordinate, UtcDatetime
from packages.schemas.weather import (
    AtmosphereData,
    DataQuality,
    DataStatus,
    WeatherForecastPoint,
    WeatherProviderInfo,
    WindVector,
)


class WeatherResponse(BaseModel):
    """Canonical API response contract for weather and wind observations."""

    location: Coordinate = Field(
        ...,
        description="Geographic coordinate of observation",
    )
    observed_at: UtcDatetime = Field(
        ...,
        description="Observation / model run epoch",
    )
    retrieved_at: UtcDatetime = Field(
        ...,
        description="Time data was fetched by the system",
    )
    data_status: DataStatus = Field(
        ...,
        description="Freshness status (LIVE, CACHED, UNAVAILABLE)",
    )
    data_quality: DataQuality = Field(
        default=DataQuality.LIVE,
        description="Quality assurance status (LIVE, CACHED, FALLBACK, UNAVAILABLE)",
    )
    atmosphere: AtmosphereData = Field(
        ...,
        description="Atmospheric conditions",
    )
    wind: WindVector = Field(
        ...,
        description="Wind speed, directional angles, cardinal labels, and decomposed vector",
    )
    forecast: list[WeatherForecastPoint] = Field(
        default_factory=list,
        description="Short-term forecast trajectory points",
    )
    provider: WeatherProviderInfo = Field(
        ...,
        description="Meteorological provider provenance",
    )


class EventWeatherResponse(BaseModel):
    """Event-coupled meteorological context response."""

    event_id: str = Field(
        ...,
        description="Canonical thermal anomaly event identifier",
    )
    weather: WeatherResponse = Field(
        ...,
        description="Meteorological conditions at the event coordinate",
    )
    enriched_at: UtcDatetime = Field(
        ...,
        description="Timestamp when event was coupled with weather context",
    )
